File: otel-extension/src/main/scala/io/scalac/mesmer/otelextension/instrumentations/akka/actor/impl/LocalActorRefProviderAdvice.scala
package io.scalac.mesmer.otelextension.instrumentations.akka.actor.impl

import akka.actor.ActorRef
import akka.actor.ActorSystem
import akka.actor.typed.scaladsl.adapter.ClassicActorSystemOps
import net.bytebuddy.asm.Advice.Argument
import net.bytebuddy.asm.Advice.OnMethodExit
import net.bytebuddy.asm.Advice.Return

import io.scalac.mesmer.core.event.ActorEvent.ActorCreated
import io.scalac.mesmer.core.event.EventBus
import io.scalac.mesmer.core.model.ActorRefTags

object LocalActorRefProviderAdvice {

  @OnMethodExit
  def actorOf(@Return ref: ActorRef, @Argument(0) system: ActorSystem): Unit =
    EventBus(system.toTyped)
      .publishEvent(ActorCreated(ActorRefTags(ref, Set.empty)))
}

File: otel-extension/src/main/java/akka/actor/impl/LocalActorRefProviderAdvice.java
package akka.actor.impl;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import io.scalac.mesmer.otelextension.instrumentations.akka.actor.ActorLifecycleEvent;
import net.bytebuddy.asm.Advice;

public class LocalActorRefProviderAdvice {

  @Advice.OnMethodExit
  public static void actorOfExit(
      @Advice.Argument(0) ActorSystem system, @Advice.Return ActorRef ref) {
    system.eventStream().publish(new ActorLifecycleEvent.ActorCreated(ref));
  }
}
